eth/fetcher/tx_fetcher.go

```diff
@@ -0,0 +1,507 @@
+				delivered := make(map[common.Hash]struct{})
+				for _, hash := range delivery.hashes {
+					delivered[hash] = struct{}{}
+				}
+				var cutoff int
+				for i, hash := range req.hashes {
+					if _, ok := delivered[hash]; ok {
+						cutoff = i
+					}
+				}
+				// Reschedule missing hashes from alternates, not-fulfilled from alt+self
+				for i, hash := range req.hashes {
+					if _, ok := delivered[hash]; !ok {
+						if i > cutoff {
+							f.alternates[hash][delivery.origin] = struct{}{}
+						}
+					}
+					if _, ok := f.announced[hash]; ok {
+						panic("announced tracker already contains alternate item")
+					}
+					f.announced[hash] = f.alternates[hash]
+					delete(f.alternates, hash)
+					delete(f.fetching, hash)
+				}
+				// Something was delivered, try to rechedule requests
+				f.scheduleFetches(timeoutTimer, map[string]struct{}{delivery.origin: struct{}{}})
+			}
+			if f.cleanupHook != nil {
+				f.cleanupHook(delivery.hashes)
+			}
+
+		case <-f.quit:
+			return
+		}
+	}
+}
+
+// rescheduleWait iterates over all the transactions currently in the waitlist
+// and schedules the movement into the fetcher for the earliest.
+func (f *TxFetcher) rescheduleWait(timer *time.Timer) {
+	if len(f.waittime) == 0 {
+		return
+	}
+	earliest := time.Now()
+	for _, instance := range f.waittime {
+		if earliest.After(instance) {
+			earliest = instance
+		}
+	}
+	timer.Reset(arriveTimeout - time.Since(earliest))
+}
+
+// rescheduleTimeout iterates over all the transactions currently in flight and
+// schedules a cleanup run when the first would trigger.
+func (f *TxFetcher) rescheduleTimeout(timer *time.Timer) {
+	if len(f.waittime) == 0 {
+		return
+	}
+	earliest := time.Now()
+	for _, req := range f.requests {
+		if earliest.After(req.time) {
+			earliest = req.time
+		}
+	}
+	timer.Reset(fetchTimeout - time.Since(earliest))
+}
+
+// scheduleFetches starts a batch of retrievals for all available idle peers.
+func (f *TxFetcher) scheduleFetches(timeout *time.Timer, whitelist map[string]struct{}) {
+	// Gather the set of peers we want to retrieve from (default to all)
+	actives := whitelist
+	if actives == nil {
+		actives = make(map[string]struct{})
+		for peer := range f.announces {
+			actives[peer] = struct{}{}
+		}
+	}
+	// For each active peer, try to schedule some transaction fetches
+	idle := len(f.requests) == 0
+	for peer := range actives {
+		if f.requests[peer] == nil {
+			// Found an idle peer, gather up a batch of hashes to fetch
+			hashes := make([]common.Hash, 0, maxTransactionFetch)
+			for hash := range f.announces[peer] {
+				if _, ok := f.fetching[hash]; !ok {
+					hashes = append(hashes, hash)
+					if len(hashes) >= maxTransactionFetch {
+						break
+					}
+				}
+			}
+			// If any hashes were allocated, mark those as fetching
+			if len(hashes) > 0 {
+				for _, hash := range hashes {
+					if _, ok := f.fetching[hash]; ok {
+						panic("fetching tracker already contains fetching item")
+					}
```

Comment:
Can't you simply delete this? Why loop twice here?

Reply:
Indeed, we construct the hashes based on fetching not containing them, so this panic is a noop.

Reply:
I meant delete the entire segment between L471:482, basically (yes, including the panic, and obviously keep the exit-condition). But mash it into one loop